### DISCUSSION TITLE
x-data with fetch, display result in x-text

### DISCUSSION THREAD

**Discussion by madsem** [alpinejs/alpine#2643] [1. Help] [answered]
Hey There,

I'm pretty new to AlpineJS, trying to make dynamic text insertion work but I get stuck.

What I'm trying to achieve: the `x-data`  directive should query the `dynamicInsertion` function, which in turn uses `fetch` to get results from a Netlify serverless function. Then the `x-text` directive should display the result.

I'm not sure what I can do in the `window.dynamicInsertion` to make it work, it always returns the promise itself not the expected string.

What I have so far:
```html
<div x-data="{ dki: dynamicInsertion('key', 'Fallback Text') }"><span x-text="dki"></span></div>
```

```js
import fetch from "unfetch";

window.dynamicInsertion = (key, fallback) => {

  fetch("/.netlify/functions/dynamic-insertion", {
    method: "POST",
    body: JSON.stringify({
      param: key,
    }),
  })
  .then((res) => res.json())
  .then((json) => {
    return {
      dki: json
    }
  })
  .catch((error) => {
    return {
      dki: fallback
    }
  });
};
```

Netlify function (not complete, just test setup):
```js
exports.handler = async function(event, context) {

  const { param } = JSON.parse(event.body);

  return {
    statusCode: 200,
    body: JSON.stringify('San Francisco')
  }

}
```

**Comment by SimoTod** [comment, marked as answer]
https://alpinejs.dev/advanced/async

**Reply by madsem** [reply]
Thanks @SimoTod but it still spits out `[object Promise]` on the page, or it's just empty

Changed the function to be async 

```html
# tried:
<div x-data="{ dki: dynamicInsertion('key', 'Fallback Text') }"><span x-text="dki"></span></div>
<div x-data="{ dki: await dynamicInsertion('key', 'Fallback Text') }"><span x-text="dki"></span></div>
<div><span x-text="dynamicInsertion('key', 'Fallback Text')"></span></div>
```

```js
import fetch from "unfetch";

window.dynamicInsertion = async (key, fallback) => {

  let re = await fetch("/.netlify/functions/dynamic-insertion", {
    method: "POST",
    body: JSON.stringify({
      param: key,
    }),
  });

  //console.log(await re.json())

  return await re.json()
};

```

**Reply by SimoTod** [reply]
you can't do it directly in x-data in that way. You can set it to a default empty value and initialise it in x-init: https://codepen.io/SimoTod/pen/RwjGJVv

**Reply by madsem** [reply]
Ahhh, gotcha @SimoTod! Thanks a lot, it all makes sense now :)
